Convert `!` commands to newshell

wargio opened the pull request:
# squash me

 <!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

- ✅ `!                        list all historic commands`
  - Moved under `H`
- ✅ `!ls                      execute 'ls' in shell`
- ✅ `!echo $RZ_SIZE           display file size`
  - **Removed** `RZ_BYTES` and `RZ_BLOCK` because useless
- ❌ `!*rzp x                  run rizin command via rzpipe in current session`
  - **Removed**
- ✅ `!!                       save command history to hist file`
  - Moved under `H+`
- ✅ `!!ls~txt                 print output of 'ls' and grep for 'txt'`
- ❌ `!!!cmd [args|$type]      adds the autocomplete value`
  - **Removed**
- ❌ `!!!-cmd [args]           removes the autocomplete value`
  - **Removed**
- ✅ `!-                       clear history in current session`
  - **Moved under** `H-`
- ❌ `!-*                      clear and save empty history log`
  - **Removed**

**Issues**

- Fix https://github.com/rizinorg/rizin/issues/1134

wargio committed:
Updated core_plugin_example.c

wargio committed:
Removed RZ_BYTES and RZ_BLOCK

These 2 variables were not used and badly implemented, therefore removed

wargio committed:
Implemented !, !<cmd>, !- and !-*

!      Shows history
!-     Clears history
!-*    Clears history and saves it
!<cmd> Executes a command on the shell.

wargio committed:
removed !-* and added !+ (!! is now deprecated)

Originally !-* was used to clear the history and save it, but makes no sense having it.
Originally !! was used to save the history.
For completeness, since !- clears, !+ shall save history

wargio committed:
Implemented !!<cmd> [args]

!! Runs a given commands in system(3) and pipes stdout to rizin

wargio committed:
Moved list history under !=

It was a bit confusing having !<cmd> and !<num> under the same command,
therefore i moved the show list history and execute history N under !=

XVilka commented on librz/core/cfile.c:
```diff
@@ -561,16 +529,14 @@ CURSOR cursor position(offset from curseek)
-	rz_sys_setenv("RZ_DEBUG", rz_config_get_b(core->config, "cfg.debug") ? "1" : "0");
-	rz_sys_setenv("RZ_IOVA", rz_config_get_i(core->config, "io.va") ? "1" : "0");
+	rz_sys_setenv("RZ_COLOR", sdb_fmt("%" PFMT64u, rz_config_get_i(core->config, "scr.color")));
+	rz_sys_setenv("RZ_DEBUG", rz_config_get_b(core->config, "cfg.debug") ? "true" : "false");
```
I suggest to just create some uber-simple RzUtil function to convert `bool` to string and use it here.

wargio replied:
tbh, i do not see any reason to have this. maybe a macro, but still, not really needed.

XVilka replied:
In fact there is already: `rz_str_bool()`. Moreover, if you just use `rz_config_get(core->config, "cfg.debug")` it should return "true" or "false" string - it checks if the variable is boolean internally already, see `librz/config/config.c`.

wargio replied:
that code is old. i will revert it back to the original code since i dropped due the new usage of the RzSubprocess

XVilka commented on librz/core/p/core_plugin_example.c:
```diff
@@ -1,40 +1,85 @@
+// SPDX-FileCopyrightText: 2021 somebody <somebody@domain.tld>
 // SPDX-License-Identifier: LGPL-3.0-only
 #if 0
```
Can you convert this comment into Doxygen `\file` instead?

XVilka reviewed (changes requested):
Overall looks good.

ret2libc commented on librz/core/cmd_descs/cmd_system.yaml:
```diff
@@ -0,0 +1,71 @@
+# SPDX-FileCopyrightText: 2021 RizinOrg <info@rizin.re>
+# SPDX-License-Identifier: LGPL-3.0-only
+---
+name: cmd_system3
```
why `3`?

XVilka replied:
Agree. Should be `42`.

wargio replied:
because is the cmd that calls system?

ret2libc commented on librz/core/cmd_descs/cmd_system.yaml:
```diff
@@ -0,0 +1,71 @@
+            arg_str: "ls~txt"
+            comment: "executes the 'ls' command via system(3) and grep for 'txt'"
+  # These commands should me moved from ! to somewhere else.
+  - name: "!="
```
where did you get these from?

wargio replied:
It was a bit confusing having `!<cmd>` and `!<num>` under the same command,
therefore i moved the `show list history` and `execute history N` under `!=`

Btw, I added all my reasoning in my commit message; you can click on the 3 dots and read the body message
![image](https://user-images.githubusercontent.com/561184/122897367-3aa71080-d34a-11eb-9bcc-533bc377a7a1.png)

ret2libc replied:
:+1:

wargio committed:
Renamed system3 to system

XVilka commented on librz/core/cmd_system.c:
```diff
@@ -0,0 +1,123 @@
+}
+
+RZ_IPI RzCmdStatus rz_clear_history_handler(RzCore *core, int argc, const char **argv) {
+	if (argc != 1) {
```
I think the arguments count now checked by the generic implementation and now unnecessary if they are specified as non-optional in YAML. @ret2libc might correct me on that.

wargio replied:
if it is, i will remove this.

ret2libc replied:
yes it is unnecessary.

ret2libc commented on librz/core/cmd_system.c:
```diff
@@ -0,0 +1,123 @@
+}
+
+RZ_IPI RzCmdStatus rz_system_handler(RzCore *core, int argc, const char **argv) {
+	if (argc == 1) {
+		rz_line_hist_list();
+		return RZ_CMD_STATUS_OK;
+	}
```
Is this still needed?

wargio replied:
i forgot to remove this.

ret2libc commented on librz/core/cmd_system.c:
```diff
@@ -0,0 +1,123 @@
+}
+
+RZ_API RzCmdStatus rz_system_to_cons_handler(RzCore *core, int argc, const char **argv) {
+	if (argc < 2) {
+		return RZ_CMD_STATUS_WRONG_ARGS;
+	}
+
```
not necessary

ret2libc commented on librz/core/cmd_system.c:
```diff
@@ -0,0 +1,123 @@
+	if (argc == 1) {
+		rz_line_hist_list();
+		return RZ_CMD_STATUS_OK;
+	} else if (argc > 2) {
+		return RZ_CMD_STATUS_WRONG_ARGS;
```
not necessary

ret2libc commented on librz/core/cmd_system.c:
```diff
@@ -0,0 +1,123 @@
+}
+
+RZ_IPI RzCmdStatus rz_history_save_handler(RzCore *core, int argc, const char **argv) {
+	if (argc != 1) {
+		return RZ_CMD_STATUS_WRONG_ARGS;
+	}
+
```
same

ret2libc commented on librz/core/cmd_system.c:
```diff
@@ -0,0 +1,123 @@
+static char *system_cmd_new(int argc, const char **argv) {
+	if (argc == 1) {
+		return strdup(argv[0]);
+	}
+
+	RzStrBuf *sb = rz_strbuf_new(argv[0]);
+	if (!sb) {
+		return NULL;
+	}
+
+	for (int i = 1; i < argc; ++i) {
+		rz_strbuf_append(sb, " \"");
+		const char *arg = argv[i];
+		if (strchr(arg, '"') || strchr(arg, '\\')) {
+			size_t p, s, len = strlen(arg);
+			for (p = 0, s = 0; p < len; ++p) {
+				if (arg[p] == '\\' || arg[p] == '"') {
+					rz_strbuf_append_n(sb, arg + s, p - s);
+					rz_strbuf_append(sb, "\\");
+					s = p;
+				}
+			}
+			arg += s;
+		}
+		rz_strbuf_appendf(sb, "%s\"", arg);
+	}
+
+	return rz_strbuf_drain(sb);
+}
```
It remains to see whether this works well on windows... I remember that I wanted to switch all this `!` stuff to RzSubprocess API.

wargio replied:
I can add a comment to refactor this later.

wargio committed:
Removed old history commands and simplified !

wargio committed:
History commands moved under H

wargio committed:
Added doxygen comment for core plugin

wargio committed:
Converted |, ! and !! to use RzSubprocess

Removed XOFFSET since it is a dup
Fixed oob read in RzSubprocess
Fixed all the tests to use the proper piping method.
Fixed issue with internal env vars to not be applied due shell not being called.

wargio commented:
```
E: Failed to fetch http://apt.llvm.org/focal/dists/llvm-toolchain-focal-12/main/binary-amd64/Packages  File has unexpected size (9437 != 9427). Mirror sync in progress? [IP: 151.101.250.49 443]
```

XVilka commented on test/db/tools/rz:
```diff
@@ -98,6 +98,7 @@ EOF
 RUN
 
 NAME=no-nl-at-eof script
+BROKEN=1
```
Why?

wargio replied:
because it does work on my linux but not on the linux machine via the CI and i have no clue why

XVilka reviewed (approved):
Overall is good to merge even as is. Just a couple minor things.

ret2libc commented on librz/core/cmd_descs/cmd_history.yaml:
```diff
@@ -0,0 +1,29 @@
+name: cmd_history
+commands:
+  - name: "H"
+    cname: list_or_exec_history
```
i prefer `history_list_or_exec`, `history_clear` and `history_save`. we usually put the "topic" as first word

ret2libc commented on librz/core/cmd_history.c:
```diff
@@ -0,0 +1,31 @@
+	int index = atoi(argv[1]);
+	if (index < 1) {
+		RZ_LOG_ERROR("index must be a positive number.\n");
+		return RZ_CMD_STATUS_WRONG_ARGS;
```
Mmm maybe STATUS_ERROR even in this case? This is no different than passing a too big number.

ret2libc commented on librz/core/cmd_system.c:
```diff
@@ -0,0 +1,92 @@
+		return NULL;
+	}
+
+	for (int i = 1; i < argc; ++i) {
+		const char *arg = argv[i];
+		size_t len = strlen(arg);
+		if (len < 1) {
+			rz_strbuf_append(sb, " \"\"");
+			continue;
+		} else if (!has_spaces(arg) || has_quotes(arg, len)) {
+			rz_strbuf_appendf(sb, " %s", arg);
+			continue;
+		}
+		rz_strbuf_append(sb, " \"");
+		if (strchr(arg, '"') || strchr(arg, '\\')) {
+			size_t p = 0, s = 0;
+			for (p = 0, s = 0; p < len; ++p) {
+				if (arg[p] == '\\' || arg[p] == '"') {
+					rz_strbuf_append_n(sb, arg + s, p - s);
+					rz_strbuf_append(sb, "\\");
+					s = p;
+				}
+			}
+			arg += s;
+		}
+		rz_strbuf_appendf(sb, "%s\"", arg);
+	}
```
not a big fan of all this logic actually... I *think* @thestr4ng3r already did something like this somewhere. Am I wrong Florian?

ret2libc commented on librz/core/cmd_system.c:
```diff
@@ -0,0 +1,69 @@
+	}
+
+	// wait max 1 min
+	RzSubprocessWaitReason reason = rz_subprocess_wait(proc, 60000);
```
Why the limit? If you do !vim you don't expect it to be closed after one min

XVilka commented:
Overall is fine, but could you please run it on more CI configurations, including older CentOS, Debian, Alpine by sending the PR from the `dist-` branch instead?

XVilka commented on librz/core/cfile.c:
```diff
@@ -492,19 +492,8 @@ RZ_API bool rz_core_file_resize_delta(RzCore *core, st64 delta) {
-			free(f);
-		}
-	}
+RZ_API void rz_core_sysenv_end(RzCore *core) {
```
Since you modify this function anyway, could you please add a Doxygen description here?

wargio replied:
dropping support for this.

XVilka commented on librz/core/cfile.c:
```diff
@@ -516,33 +505,12 @@ RZ_API void rz_core_sysenv_end(RzCore *core, const char *cmd) {
-
-	RZ_API char *rz_core_sysenv_begin(RzCore *core, const char *cmd) {
-	char *f, *ret = cmd ? strdup(cmd) : NULL;
+RZ_API void rz_core_sysenv_begin(RzCore *core) {
```
Sam (was) here

wargio replied:
same. dropping support.

XVilka commented on librz/core/cmd.c:
```diff
@@ -2087,7 +2081,7 @@ static void rz_w32_cmd_pipe(RzCore *core, char *rizin_cmd, char *shell_cmd) {
 #endif
 
-RZ_API int rz_core_cmd_pipe(RzCore *core, char *rizin_cmd, char *shell_cmd) {
+RZ_API int rz_core_cmd_pipe_old(RzCore *core, char *rizin_cmd, char *shell_cmd) {
```
Could you please add Doxygen description here?

wargio replied:
not going to happen, since it is old code and likely being removed once the code does work.

XVilka commented on librz/core/cmd.c:
```diff
@@ -2163,6 +2157,59 @@ RZ_API int rz_core_cmd_pipe(RzCore *core, char *rizin_cmd, char *shell_cmd) {
+	return output;
+}
+
+RZ_API RzCmdStatus rz_core_cmd_pipe(RzCore *core, char *rizin_cmd, int argc, char **argv) {
```
Since this is new API function, add Doxygen please.

wargio replied:
not sure if we should support this as an API.

XVilka replied:
Probably not.

ret2libc replied:
please anyway a small description would be good IMHO.

wargio committed:
Fixed some tests.

wargio committed:
Use raw output when handling pipes.

Updated grammar following @ret2libc comments

wargio committed:
Fixed linter for grammar.js

wargio committed:
Expect string if no length is required.

wargio committed:
Always add null byte to the end of the buffer.

wargio committed:
Fixed rz_find script

ret2libc commented on librz/include/rz_util/rz_subprocess.h:
```diff
@@ -72,8 +72,8 @@ RZ_API void rz_subprocess_free(RzSubprocess *proc);
 RZ_API int rz_subprocess_ret(RzSubprocess *proc);
-RZ_API char *rz_subprocess_out(RzSubprocess *proc);
-RZ_API char *rz_subprocess_err(RzSubprocess *proc);
+RZ_API char *rz_subprocess_out(RzSubprocess *proc, int *length);
```
what about
```suggestion
RZ_API char *rz_subprocess_out(RzSubprocess *proc, size_t *length);
```
?

wargio replied:
i thought about this, but RzStrbuf does not use size_t.

ret2libc replied:
:/ ok

ret2libc commented on librz/util/subprocess.c:
```diff
@@ -632,7 +620,7 @@ static char **create_child_env(const char *envvars[], const char *envvals[], siz
 		new_env[i] = strdup(environ[i]);
 	}
-	for (size_t i = 0; i <= env_size; i++) {
+	for (size_t i = 0; i <= new_env_size; i++) {
```
nice catch... I guess it was wrong before.

wargio replied:
i had a huge headache on this since asan was complaining but not telling me where due optimizations.

ret2libc replied:
compile with `-buildtype=debug` if that was the problem.

ret2libc commented on test/db/tools/rz_bin:
```diff
@@ -155,87 +155,87 @@ RUN
 NAME=rz-bin -zzz pe
 FILE=bins/pe/ioli/w32/crackme0x00.exe
-CMDS=!rz-bin -zzz ${RZ_FILE} | grep "Password:"
+CMDS=!!rz-bin -zzz ${RZ_FILE} | grep "Password:"
```
I guess this kind of change is now required because the `|` is handled by the rz-shell, so you have to redirect the output/err to rzcons, so that it can be later sent to `grep`, right?

wargio replied:
Yes. because !! does fetch stdout, and ! does not

ret2libc commented on librz/core/cfile.c:
```diff
@@ -561,16 +529,14 @@ CURSOR cursor position(offset from curseek)
-	rz_sys_setenv("RZ_COLOR", rz_config_get_i(core->config, "scr.color") ? "1" : "0");
-	rz_sys_setenv("RZ_DEBUG", rz_config_get_b(core->config, "cfg.debug") ? "1" : "0");
-	rz_sys_setenv("RZ_IOVA", rz_config_get_i(core->config, "io.va") ? "1" : "0");
+	rz_sys_setenv("RZ_COLOR", sdb_fmt("%" PFMT64u, rz_config_get_i(core->config, "scr.color")));
```
`rz_strf` please.

wargio replied:
i'll reset this back to the original code and even maybe remove this.

ret2libc commented on librz/core/cmd.c:
```diff
@@ -2163,6 +2157,59 @@ RZ_API int rz_core_cmd_pipe(RzCore *core, char *rizin_cmd, char *shell_cmd) {
+		.file = argv[0],
+		.args = (const char **)&argv[1],
+		.args_size = argc - 1,
+		.envvars = NULL,
```
do we want to pass some vars to the subprocess? `RZ_FILE`, `RZ_OFFSET`, `RZ_ENDIAN`, etc.?

wargio replied:
hmm i do not know, because i expect the app to eat the stdin not the envvars.

ret2libc replied:
according to `!??` programs are expected to find some env vars. It seems that in the past implementation that was all set with `rz_core_sysenv_begin`/`rz_core_sysenv_end`. It should be fine to keep this as-is, because current implementation of RzSubprocess just allows to add/replace env variables, but not to remove them. So by setting `.envvars = NULL`  we are really just keeping the parent env vars, which should be fine.

ret2libc commented on librz/core/cmd.c:
```diff
@@ -6007,17 +6054,22 @@ DEFINE_HANDLE_TS_FCN_AND_SYMBOL(html_enable_stmt) {
+
+	RzCmdStatus res = RZ_CMD_STATUS_INVALID;
+	RzCmdParsedArgs *a = ts_node_handle_arg_prargs(state, node, command_pipe, 1, true);
+	if (a && a->argc > 1) {
```
I would check `if (!a || a->argc <= 1) return INVALID`. Otherwise you just ignore the error, which I think is never good.

wargio replied:
that is not true. `res` is `RZ_CMD_STATUS_INVALID` so if that `if` fails, it will return `RZ_CMD_STATUS_INVALID`

ret2libc replied:
Right!

ret2libc commented on librz/core/cmd.c:
```diff
@@ -6563,6 +6614,27 @@ RZ_API char *rz_core_cmd_str(RzCore *core, const char *cmd) {
+	ut8 *retstr = NULL;
+	rz_cons_push();
+	if (rz_core_cmd(core, cmd, 0) == -1) {
+		//eprintf ("Invalid command: %s\n", cmd);
```
`rz_cons_pop`

wargio replied:
i saw that you added this in your PR, should i skip this change ? @ret2libc

ret2libc replied:
I will just do a separate PR just with this ;) my PR will take some time to be done.

ret2libc replied:
This is still to be done.

ret2libc reviewed (commented):
Awesome work @wargio , thanks a lot for moving this thing forward! I haven't tested it yet, I will do after some more changes, however things look nice so far!

wargio committed:
Refactoring of history commands

wargio committed:
Added comments on rz_sysenv to be deprecated

wargio committed:
Removed old command rz_cmd_system

wargio committed:
Comments on new apis

XVilka reviewed (approved):
LGTM. But please don't forget to test on `dist-` branch before merging.

XVilka commented:
94 broken tests: https://builds.sr.ht/~xvilka/job/534811#task-test-0

wargio committed:
Fixed windows pipes and stdin/out/err

wargio committed:
Fixed env vars in command line

wargio committed:
Allow to have null bytes on stdout/err

wargio committed:
removed !!! code

wargio committed:
Added missing definition

wargio committed:
Small fixes

wargio committed:
Removed old code

wargio committed:
Fixed unit test

wargio committed:
remove cr

wargio committed:
fixing rzpipe

wargio committed:
Fixed missing header

wargio committed:
Fixed rz_asm test - no double quotes.

wargio committed:
Removed useless double quotes.

wargio committed:
Fixed mangling test - wrong double quote usage

wargio committed:
Fixing CR issue on windows

wargio committed:
Always remove CR on windows results

wargio committed:
Merge branch 'dev' into shell-cmd-1

wargio committed:
Merge branch 'dev' into shell-cmd-1

wargio committed:
Merge branch 'dev' into shell-cmd-1

wargio committed:
Merge branch 'dev' into shell-cmd-1

wargio committed:
Merge branch 'dev' into shell-cmd-1

wargio committed:
Merge branch 'dev' into shell-cmd-1

ret2libc commented on librz/core/cfile.c:
```diff
@@ -563,16 +533,14 @@ CURSOR cursor position(offset from curseek)
-	rz_sys_setenv("RZ_DEBUG", rz_config_get_b(core->config, "cfg.debug") ? "1" : "0");
-	rz_sys_setenv("RZ_IOVA", rz_config_get_i(core->config, "io.va") ? "1" : "0");
+	rz_sys_setenv("RZ_BITS", rz_config_get(core->config, "asm.bits"));
+	rz_sys_setenv("RZ_COLOR", rz_config_get(core->config, "scr.color"));
+	rz_sys_setenv("RZ_DEBUG", rz_config_get(core->config, "cfg.debug"));
+	rz_sys_setenv("RZ_IOVA", rz_config_get(core->config, "io.va"));
```
This is not exactly the same as before.. For example, for booleans vars it would returne `true`/`false` instead of `1`/`0` as before. I think for env vars it is more common to have 1/0 ?

wargio replied:
no. this better and this stuff is not used besides the user itself.

ret2libc replied:
How is this better exactly? Anyway, I'm ok with having `true`/`false` instead of 1/0, but please update the documentation because it says 1/0.

wargio replied:
book needs to be updated anyway after this PR

ret2libc replied:
I meant the rizin doc for `%??`.

By the way, I changed my mind. I think there is no need to do this change here and now. In particular, there are places that seem to assume boolean env vars as 1/0 and not true/false,  so I'll revert these.

ret2libc commented on librz/core/cmd_history.c:
```diff
@@ -0,0 +1,31 @@
+		return RZ_CMD_STATUS_OK;
+	}
+
+	int index = atoi(argv[1]);
```
`rz_num`?

wargio replied:
no. i expect a number and i want a number. atoi is enough

ret2libc commented on librz/core/cmd_system.c:
```diff
@@ -0,0 +1,217 @@
+	const char *scr_color = rz_config_get(core->config, "scr.color");
+	const char *endian = rz_str_bool(core->rasm->big_endian);
+	char *cfg_path = config_path(core);
+	snprintf(file_size, sizeof(file_size), "%" PFMT64u, core->file ? rz_io_fd_size(core->io, core->file->fd) : 0);
+	snprintf(core_offset, sizeof(core_offset), "%" PFMT64u, core->offset);
+	snprintf(block_size, sizeof(block_size), "%u", core->blocksize);
```
```suggestion
	rz_strf(file_size, "%" PFMT64u, core->file ? rz_io_fd_size(core->io, core->file->fd) : 0);
	rz_strf(core_offset, "%" PFMT64u, core->offset);
	rz_strf(block_size, "%u", core->blocksize);
```

ret2libc commented on librz/core/cmd_system.c:
```diff
@@ -0,0 +1,217 @@
+		scr_color,
+		block_size,
+		cfg_debug,
+		cfg_path ? cfg_path : ""
```
```suggestion
		rz_str_get(cfg_path)
```

ret2libc commented on librz/core/cmd_system.c:
```diff
@@ -0,0 +1,217 @@
+	}
+
+	if (!rz_subprocess_init()) {
+		RZ_LOG_ERROR("Cannot initialize subprocess.\n");
+		rz_list_free(alloc);
+		rz_file_rm(cfg_path);
+		free(cfg_path);
+		free(args);
+		return false;
```
put this at the end of the function and use `goto`, it will simplify errors handling and make sure things are done in the right order on each error.

ret2libc commented on librz/core/cmd_system.c:
```diff
@@ -0,0 +1,217 @@
+		return value;
+	}
+
+	if ((it = rz_list_find_ptr(alloc, (void *)arg))) {
+		string = it->data;
+	} else {
+		string = strdup(arg);
+	}
+
```
I don't see what advantages `alloc` array would bring. I would just go with `rz_str_replace` and remove `alloc`. Moreover, you can only find the same `arg` in `alloc` only if it's the same argv[i], so probably a list is not really necessary anyway.

wargio replied:
alloc, is just a simple idea where i place the allocated strings and just free them via rz_list so i am sure that i never forget to free an allocated value, so i do not have to care whether argv is allocated on the heap or not, using directly rz_str_replace is a super bad idea since you do not know where those arguments are really allocated.

ret2libc reviewed (commented):
Just looked at some parts for now! Doing another round in a bit.

ret2libc commented on librz/core/cmd.c:
```diff
@@ -2163,6 +1925,66 @@ RZ_API int rz_core_cmd_pipe(RzCore *core, char *rizin_cmd, char *shell_cmd) {
+ * Executes a rizin command and the raw bytes of the stdout is returned.
+ * The output is then used as stdin for execvp.
+ * The output of the execvp is then sent into RzCons.
+ * */
```
This should be better explained. In particular, it is not clear what is piped where. In this case, `rizin_cmd` is the first command and its output is piped as stdin of the command specified in (argc, argv).

ret2libc commented on librz/core/cmd.c:
```diff
@@ -6565,6 +6391,32 @@ RZ_API char *rz_core_cmd_str(RzCore *core, const char *cmd) {
+ *
+ * Executes a rizin command and returns the raw stdout and its length.
+ * */
+RZ_API ut8 *rz_core_cmd_raw(RzCore *core, const char *cmd, int *length) {
```
Probably this can be unified with rz_core_cmd_str to avoid code dup.

ret2libc commented on librz/core/p/core_plugin_example.c:
```diff
@@ -1,41 +1,93 @@
-mkdir -p ~/.config/rizin/plugins
-mv core_plugin_example.so ~/.config/rizin/plugins
-#endif
+
```
Awesome! Is this compiled anyhow while testing rizin? I was also wondering.. Aren't unit tests enough to showcase how to write a new command?

wargio replied:
yes

ret2libc replied:
Where is this compiled?

wargio replied:
on my machine.

ret2libc commented on test/db/io/baddr:
```diff
@@ -17,7 +17,7 @@ RUN
 NAME=3: io.baddr
 FILE=bins/elf/analysis/pie
-CMDS=!rz-bin -B 0xf00000 -re ${RZ_FILE} | grep 0x
+CMDS=!!rz-bin -B 0xf00000 -re ${RZ_FILE} | grep 0x
```
I think this should be kinda automatic or, at least, the default behaviour, without having to use `!!` all the time.
If we read the help of `|` it is `| <cmd> | <program> # Pipe output of command to program`, so by using `|` the user already express his will to pipe the output of one "command" to another program. 

Do you think it would make sense to detect the `pipe` and assume output should be redirected to rzcons?

wargio replied:
i disagree
you can't know if this is piped, or not

ret2libc replied:
Why can't you? In `pipe_stmt` handler you know if a program is piped or not.

wargio replied:
yes, but what if my cmd is like 
```
[0x00000000] `aaa | bbb` | ccc
```

ret2libc replied:
Not sure about it's better `!!<prog1> | <prog2>` or `!<prog1> | <prog2>`...

Choosing `!!<prog1> | <prog2>` means that essentially you never use `!<prog1> | <prog2>` and you have to type `!!` every time. Also, as a user you first do `!ls` and then you probably want to grep it and suddenly you have to do `!!ls` instead of just `!ls` if you want to pipe. On the other hand, this maintains the semantic of `!` and `!!`.

Choosing `!<prog1> | <prog2>` and make sure that `!<prog1>` automatically redirects stdout to rzcons whenever a pipe is used means that as a user it's easier to construct a command without really having to *know* about rzcons (which is a very internal thing after all..). However, it loses a bit the semantic of `!` vs `!!`.

WDYT @wargio  @XVilka @thestr4ng3r ?

wargio replied:
for me you can remove `!!` and merge them into `!` and do the autodetect for pipe/grep/etc..

XVilka replied:
I think autodetection will provide the best user experience, indeed

ret2libc commented on librz/util/subprocess.c:
```diff
@@ -36,13 +36,17 @@ static bool create_pipe_overlap(HANDLE *pipe_read, HANDLE *pipe_write, LPSECURIT
 		return FALSE;
 	}
+
+	SetEnvironmentVariable(TEXT("RZ_PIPE_PATH"), name);
```
Why is this necessary? What does it do?

wargio replied:
used by rz-pipe on windows.

ret2libc replied:
I see :/ Does linux needs it too?

wargio replied:
as far i know, nope.

ret2libc commented on librz/util/subprocess.c:
```diff
@@ -132,29 +136,29 @@ static LPWCH override_env(const char *envvars[], const char *envvals[], size_t e
 		return NULL;
 	}
-	argv[0] = (char *)opt->file;
+	argv[0] = ""; // a space is required to work correctly.
```
why the `cmd_exe` is not placed in `argv[0]`?

wargio replied:
because we are calling `CreateProcessA` properly. check MS guide.

XVilka replied:
`""` is not a "space" though, it's an empty string.

ret2libc commented on librz/core/cmd.c:
```diff
@@ -6008,17 +5830,22 @@ DEFINE_HANDLE_TS_FCN_AND_SYMBOL(html_enable_stmt) {
+	TSNode command_rizin = ts_node_named_child(node, 0);
+	TSNode command_pipe = ts_node_named_child(node, 1);
+
+	char *rz_cmd = ts_node_sub_string(command_rizin, state->input);
```
This is probably unnecessary. We can probably use `handle_ts_stmt` as done elsewhere to direclty work with the TSNode, so we avoid going again through the whole tree-sitter parsing and handling.

ret2libc committed:
Revert "Fixed unit test"

This reverts commit bca6a588a6bb1cfd12959c35d6986beb033ce0bb.

ret2libc committed:
Revert "Removed old code"

This reverts commit 01ecb483977c1ba76132fbf81a328ceb43cc4dbe.

ret2libc committed:
Revert "removed !!! code"

This reverts commit 3d3cf42ba79bd08aaa52e794914231b4f5e1da0e.

ret2libc committed:
Revert "Small fixes"

This reverts commit f66df84cf28e5070884fad7b23a1020eaa37bacb.

ret2libc committed:
Remove unused functions/vars

ret2libc commented on test/unit/test_rzpipe.c:
```diff
@@ -11,7 +11,7 @@ static bool test_rzpipe(void) {
 	mu_assert("rzpipe can spawn", r);
 	char *hello = rzpipe_cmd(r, "?e hello world");
-	mu_assert_streq(hello, "hello world\n", "rzpipe hello world");
+	mu_assert_streq(hello, "[0x00000000]> hello world\n", "rzpipe hello world");
```
I think this change is wrong. The output of the `?e` command is just the string, not the prompt + string.

ret2libc committed:
Revert "fixing rzpipe"

This reverts commit 17cc2cdc02b63098e6f55d1042a4c62c6c843616.

ret2libc committed:
Various PR review fixes

ret2libc committed:
Try to avoid iterating the alloc list every time

ret2libc committed:
Switch back to env var with 1/0

ret2libc committed:
Fix clang-format

ret2libc committed:
Avoid calling again the whole tree-sitter pipeline

ret2libc committed:
Merge remote-tracking branch 'origin/dev' into shell-cmd-1

ret2libc commented on librz/util/subprocess.c:
```diff
@@ -132,29 +136,29 @@ static LPWCH override_env(const char *envvars[], const char *envvals[], size_t e
 	return ret;
 }
 
-static void remove_cr(char *str) {
-	char *start = str;
-	while (*str) {
-		if (str[0] == '\r' &&
-			!(str - start >= 4 && !strncmp(str - 4, RZ_CONS_CLEAR_SCREEN, 4))) {
-			memmove(str, str + 1, strlen(str + 1) + 1);
-			continue;
-		}
-		str++;
-	}
-}
-
```
@wargio I see this was removed from subprocess, but now old users of the subprocess API has to do this themselves. Did you remove this because you want to get the raw stdout/stderr of a program?

What do you think if we use these API?
```
RZ_API char *rz_subprocess_out_str(RzSubprocess *proc);
RZ_API char *rz_subprocess_err_str(RzSubprocess *proc);
RZ_API ut8 *rz_subprocess_out(RzSubprocess *proc, int *length);
RZ_API ut8 *rz_subprocess_err(RzSubprocess *proc, int *length);
```
The `_str` APIs apply `remove_cr` before returning the strings, so we don't need to replicate this function everywhere.

Together with this change we should also change a bit the `RzSubprocessOutput` structure. Something like:
```
typedef struct rz_process_output_t {
	char *out_str; // stdout
	char *err_str; // stderr
        ut8 *out, *err;
        int out_len, err_len;
	int ret; // exit code of the process
	bool timeout;
} RzSubprocessOutput;
```
Comments?

wargio replied:
stdout needs to be raw. it's up to the user to know what there will be in the buffer.
this is the reason you can do `cat mybinfile | hexdump -C` for example

wargio replied:
The remove_cr function shall not be there because the only one complaining about this is the testsuite, therefore that is a testsuite issue not an issue related to the pipe, nor subprocess

ret2libc replied:
I agree, that's why I kept the raw stdout in the proposal above. However, other times you just want the string and you want a "cleaned" string (e.g. by running remove_cr function), so I was thinking about providing that from RzSubprocess itself instead of having every user of the API definiting their own remove_cr function.

Also, no matter what we should probably add the `out_len`/`err_len` fields.

ret2libc replied:
cc @thestr4ng3r you wrote that piece of code, right?

thestr4ng3r replied:
This `remove_cr` implementation is super inefficient so I can only assume I wrote it to only be used as part of `rz-test`, where it doesn't really matter.
If another user of `RzSubprocess` needs this, the functionality should not be provided as part of the subprocess api but as just an independent `rz_str_` function that can be called.
So for me the move to `rz-test/run.c` as it is done here seems good.

ret2libc replied:
Ok thanks for confirming @thestr4ng3r . Then I will just make sure to add `out_len`/`err_len` to `RzSubprocessOutput` and that's it.

ret2libc committed:
Make RzSubprocess deal with ut8 instead of assuming strings

ret2libc committed:
Automatically detect if a command needs to be piped to another one or not

XVilka commented on test/db/cmd/feat_redirect:
```diff
@@ -2,9 +2,9 @@ NAME="pb>x;!wc -c x"
 CMDS=<<EOF
 p8 10 > dump
-!wc -c dump|awk "{print \$1}"
+!!wc -c dump|awk "{print \$1}"
```
Does this need to be updated too then?

ret2libc replied:
Yeah all these changes can be reverted back if we are ok with the auto detection.

XVilka commented on test/db/cmd/cmd_system:
```diff
@@ -16,14 +16,14 @@ RUN
 NAME=!| grep
 FILE==
-CMDS=!echo magic|grep bob
+CMDS=!!echo magic|grep bob
```
And here

wargio commented on librz/core/cmd_system.c:
```diff
@@ -143,7 +143,7 @@ static bool system_exec(RzCore *core, int argc, const char **argv, char **output
 		.env_size = RZ_ARRAY_SIZE(envvars),
 		.stdin_pipe = RZ_SUBPROCESS_PIPE_NONE,
-		.stdout_pipe = output ? RZ_SUBPROCESS_PIPE_CREATE : RZ_SUBPROCESS_PIPE_NONE,
+		.stdout_pipe = force_output || core->is_pipe ? RZ_SUBPROCESS_PIPE_CREATE : RZ_SUBPROCESS_PIPE_NONE,
```
isn't easier to check for the output ptr?

ret2libc replied:
I was not sure it is possible to do that way, because when you execute `!<cmd>` in some cases you want the output, in others you don't want it. So from rz_system_handler you have to always pass the `output` var.

wargio replied:
yes, but the output needs to go somewhere and return it, so the output pointer MUST be non NULL.

wargio replied:
i would move the core->is_pipe check one function above, then pass the pointer accordingly to the check.
i mean, this way also works.

ret2libc committed:
Apply deroad' suggestion

XVilka commented:
```
clang-cl" "-Ibinrz/rz-test\rz-test.exe.p" "-I." "-I.." "-I..\librz\include" "-Isubprojects\sdb" "-I..\subprojects\sdb" "-Isubprojects\sdb\src" "-I..\subprojects\sdb\src" "-I..\shlr\spp" "/MD" "/nologo" "/showIncludes" "/W2" "/O2" "-fcommon" "-D__STDC__=1" "-D_CRT_DECLARE_NONSTDC_NAMES " "-D_CRT_SECURE_NO_WARNINGS" "-D_CRT_NONSTDC_NO_DEPRECATE" "-Werror=sizeof-pointer-memaccess" "/Fdbinrz/rz-test\rz-test.exe.p\run.c.pdb" /Fobinrz/rz-test/rz-test.exe.p/run.c.obj "/c" ../binrz/rz-test/run.c
../binrz/rz-test/run.c(12,10): error: 'unsigned char *' and 'char *' are not pointers to compatible types
                        !(str - start >= 4 && !strncmp(str - 4, RZ_CONS_CLEAR_SCREEN, 4))) {
                          ~~~ ^ ~~~~~
../binrz/rz-test/run.c(12,35): warning: passing 'unsigned char *' to parameter of type 'const char *' converts between pointers to integer types with different sign [-Wpointer-sign]
                        !(str - start >= 4 && !strncmp(str - 4, RZ_CONS_CLEAR_SCREEN, 4))) {
                                                       ^~~~~~~
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\ucrt\string.h(272,44): note: passing argument to parameter '_Str1' here
    _In_reads_or_z_(_MaxCount) char const* _Str1,
                                           ^
../binrz/rz-test/run.c(13,33): warning: passing 'unsigned char *' to parameter of type 'const char *' converts between pointers to integer types with different sign [-Wpointer-sign]
                        memmove(str, str + 1, strlen(str + 1) + 1);
                                                     ^~~~~~~
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\ucrt\string.h(216,24): note: passing argument to parameter '_Str' here
    _In_z_ char const* _Str
                       ^
2 warnings and 1 error generated.
[1510/1682] Linking target test/unit/auxiliary/subprocess-interactive.exe
```

ret2libc committed:
Fix ut8 vs char issue

ret2libc committed:
Use ! in tests when piping, instead of git status

XVilka commented:
```
"cl" "-Ibinrz/rz-test\rz-test.exe.p" "-I." "-I.." "-I..\librz\include" "-Isubprojects\sdb" "-I..\subprojects\sdb" "-Isubprojects\sdb\src" "-I..\subprojects\sdb\src" "-I..\shlr\spp" "/MT" "/nologo" "/showIncludes" "/W2" "/O2" "/Gw" "/Fdbinrz/rz-test\rz-test.exe.p\run.c.pdb" /Fobinrz/rz-test/rz-test.exe.p/run.c.obj "/c" ../binrz/rz-test/run.c
../binrz/rz-test/run.c(9): error C2146: syntax error: missing ')' before identifier 's'
../binrz/rz-test/run.c(9): error C2059: syntax error: ')'
[1436/1658] Compiling C object binrz/rz-test/rz-test.exe.p/rz-test.c.obj
[1437/1658] Compiling C object binrz/rz-test/rz-test.exe.p/load.c.obj
ninja: build stopped: subcommand failed.
```

XVilka committed:
Fix compilation error

XVilka commented:
```
FAILED: test/unit/test_subprocess.exe.p/test_subprocess.c.obj 
"clang-cl" "-Itest/unit\test_subprocess.exe.p" "-I." "-I.." "-I..\librz\include" "-Itest/unit" "-I..\test\unit" "-Isubprojects\sdb" "-I..\subprojects\sdb" "-Isubprojects\sdb\src" "-I..\subprojects\sdb\src" "-I..\shlr\spp" "-Ishlr" "-I..\shlr" "-I..\librz\asm\arch\include" "-I..\librz\asm\arch" "-I..\librz\asm\arch\h8300" "-I..\librz\asm\arch\hexagon" "-I..\librz\asm\arch\msp430" "-I..\librz\asm\arch\rsp" "-I..\librz\asm\arch\mcore" "-I..\librz\asm\arch\v850" "-I..\librz\asm\arch\propeller" "-I..\librz\asm\arch\ebc" "-I..\librz\asm\arch\cr16" "-I..\librz\asm\arch\8051" "-I..\librz\asm\arch\v810" "-I..\librz\asm\arch\or1k" "-I..\librz\analysis\arch\gb" "-I..\librz\analysis\arch\hexagon" "-I..\librz\bin\format\elf" "-I..\librz\bin\mangling" "-I..\librz\bin\format" "-I..\shlr\heap\include\rz_windows" "-I..\librz\type\parser" "/MD" "/nologo" "/showIncludes" "/W2" "/O2" "-fcommon" "-D__STDC__=1" "-D_CRT_DECLARE_NONSTDC_NAMES " "-D_CRT_SECURE_NO_WARNINGS" "-D_CRT_NONSTDC_NO_DEPRECATE" "-Werror=sizeof-pointer-memaccess" "/Fdtest/unit\test_subprocess.exe.p\test_subprocess.c.pdb" /Fotest/unit/test_subprocess.exe.p/test_subprocess.c.obj "/c" ../test/unit/test_subprocess.c
../test/unit/test_subprocess.c(17,10): error: 'unsigned char *' and 'char *' are not pointers to compatible types
                        !(str - start >= 4 && !strncmp(str - 4, RZ_CONS_CLEAR_SCREEN, 4))) {
                          ~~~ ^ ~~~~~
../test/unit/test_subprocess.c(17,35): warning: passing 'unsigned char *' to parameter of type 'const char *' converts between pointers to integer types with different sign [-Wpointer-sign]
                        !(str - start >= 4 && !strncmp(str - 4, RZ_CONS_CLEAR_SCREEN, 4))) {
                                                       ^~~~~~~
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\ucrt\string.h(272,44): note: passing argument to parameter '_Str1' here
    _In_reads_or_z_(_MaxCount) char const* _Str1,
                                           ^
../test/unit/test_subprocess.c(18,33): warning: passing 'unsigned char *' to parameter of type 'const char *' converts between pointers to integer types with different sign [-Wpointer-sign]
                        memmove(str, str + 1, strlen(str + 1) + 1);
                                                     ^~~~~~~
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\ucrt\string.h(216,24): note: passing argument to parameter '_Str' here
    _In_z_ char const* _Str
                       ^
2 warnings and 1 error generated.
[1659/1682] Compiling C object test/unit/test_task.exe.p/test_task.c.obj
[1660/1682] Compiling C object test/unit/test_tree.exe.p/test_tree.c.obj
ninja: build stopped: subcommand failed
```

ret2libc committed:
Fix ut8 vs char issue v2

ret2libc reviewed (approved):
To me things are in good shape right now. Good work @wargio !! Please review this as well, in particular the latest changes I did.
